testing/end_to_end_test.go

```diff
@@ -0,0 +1,236 @@
+//go:build end_to_end
+// +build end_to_end
+
+// The end-to-end test is meant to exercise every potential command of the CLI in its built form
+// It's meant as a sanity check on deployed code.
+// It is not intended to catch edge cases or weird interactions; that's the realm of unit testing.
+
+// To run:
+//    go test -v -tags end_to_end ./testing
+
+package testing
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/suite"
+)
+
+// Viper Environment Variables
+const (
+	Deployment string = "deployment"
+	Config     string = "config"
+	Dev        string = "dev"
+	Staging    string = "staging"
+	Prod       string = "prod"
+)
+
+// CLI Constants
+const (
+	TempDirSuffix   string = "cli-test"
+	CliName         string = "resim"
+	ProdEndpoint    string = "https://api.resim.ai/v1/"
+	StagingEndpoint string = "https://api.resim.io/v1/"
+)
+
+type AuthConfig struct {
+	AuthKeyProviderDomain string
+	OAuthAudience         string
+}
+
+var DevAuthConfig = AuthConfig{
+	AuthKeyProviderDomain: "https://resim-dev.us.auth0.com/",
+	OAuthAudience:         "https://api.resim.ai",
+}
+
+var ProdAuthConfig = AuthConfig{
+	AuthKeyProviderDomain: "https://resim.us.auth0.com/",
+	OAuthAudience:         "https://api.resim.ai",
+}
+
+type EndToEndTestSuite struct {
+	suite.Suite
+	CliPath  string
+	Config   AuthConfig
+	Endpoint string
+}
+
+type Flag struct {
+	Name  string
+	Value string
+}
+
+type CommandBuilder struct {
+	Command string
+	Flags   []Flag
+}
+
+type Output struct {
+	StdOut string
+	StdErr string
+}
+
+func (s *EndToEndTestSuite) TearDownSuite() {
+	fmt.Println("Cleaning up")
+	os.Remove(s.CliPath)
+}
+
+func (s *EndToEndTestSuite) SetupSuite() {
+	// Set to the dev config and deployment endpoint by default:
+	var authConfig AuthConfig = DevAuthConfig
+	var deployment string = viper.GetString(Deployment)
+	s.Endpoint = fmt.Sprintf("https://%s.api.dev.resim.io/v1", deployment)
+	switch viper.GetString(Config) {
+	case Dev:
+		if deployment == "" {
+			fmt.Fprintf(os.Stderr, "error: must set CLI_E2E_TEST_DEPLOYMENT for dev")
+			os.Exit(1)
+		}
+	case Staging:
+		// Maintain the dev auth config, but set the endpoint to staging
+		s.Endpoint = StagingEndpoint
+	case Prod:
+		authConfig = ProdAuthConfig
+		s.Endpoint = ProdEndpoint
+	default:
+		s.FailNow("Invalid config value")
+	}
+	s.Config = authConfig
+	s.CliPath = s.buildCLI()
+	// Validate the RESIM_CLIENT_ID and RESIM_CLIENT_SECRET environment variables are set
+	if !viper.IsSet("RESIM_CLIENT_ID") {
```

Comment:
Also weird to have hardcoded flag variables here.  Could import the constants and construct them?  Or is that not worth the effort?

Reply:
done